Add ability to add or subtract tuples from GridQubits, ints from LineQubits

dabacon opened the pull request:
Find myself wanting to get neighbors by direction:

```
g: GridQubit
directions = [(1,0), (-1,0), (0,1), (-1,0)]
for dir in direction:
  neighbor = g + dir
```

If this is too wonky we don't have to add it (pun intended)

cduck commented:
It may be cleaner to define an instance method like `with_offset(row_offset, col_offset) -> GridQubit`.

cduck commented on cirq/devices/grid_qubit.py:
```diff
@@ -1,157 +1 @@
    @staticmethod
    def from_proto_id(proto_id: str) -> 'GridQubit':
        row, col = proto_id.split('_')
        return GridQubit(row=int(row), col=int(col))
```
Your editor mangled the file.

Strilanc replied:
The formater isn't even upset about it! Very strange.

dabacon replied:
wowah

Strilanc commented:
I think this is fine to allow.

In the past, @maffoo has advocated for making Tuple[int, int] be one-to-one with GridQubit and int be one-to-one with LineQubit. It's the sort of thing I wish we could do without risking type safety quite so much.

Strilanc commented on cirq/devices/grid_qubit_test.py:
```diff
@@ -114,6 +114,36 @@ def test_xmon_qubit_is_adjacent():
+
+def test_grid_qubit_unsupported_add():
+    with pytest.raises(AssertionError, match='1'):
+        cirq.GridQubit(1, 1) + 1
```
Use `_ = ...` to satisfy pylint

cduck commented:
Would you allow GridQubit+GridQubit also?

Strilanc commented:
Should add `+ int` to LineQubit while you're at it.

dabacon commented:
FIxed format (^M^M^M).

Added LineQubit.

Not 100% certain cduck's suggestion isn't correct.

GridQubit + GridQubit feels a bit strange to me.

PTAL.

cduck commented on cirq/devices/grid_qubit.py:
```diff
@@ -24,6 +24,11 @@ class GridQubit(ops.Qid):
+    New GridQubits can be constructed by adding or subtracting tuples
+
+        GridQubit(2, 3) + (3, 1) produces GridQubit(5, 4)
+        GridQubit(2, 3) - (1, 2) produces GridQubit(1, 1)
```
Format it like
```python
>>> GridQubit(2, 3) + (3, 1)
GridQubit(5, 4)
```
so the new doctest can test it.

dabacon replied:
Done.  It ends up being
```
>>> GridQubit(2, 3) + (3, 1)
cirq.GridQubit(5, 4)
```
It might be nice to import cirq into doc test so this could be
```
>>> cirq.GridQubit(2, 3) + (3, 1)
cirq.GridQubit(5, 4)
```

cduck commented on cirq/devices/grid_qubit.py:
```diff
@@ -131,6 +136,29 @@ def __repr__(self):
+    def __sub__(self, other: Tuple[int, int]) -> 'GridQubit':
+        assert isinstance(other, tuple) and len(other) == 2 and all(
+            isinstance(x, int) for x in other), (
+                'Can only add tuple of length 2 to GridQubit. Was {}'.format(
```
*subtract*.

cduck commented on cirq/devices/grid_qubit_test.py:
```diff
@@ -97,7 +97,7 @@ def test_xmon_qubit_ordering():
 
 
-def test_xmon_qubit_is_adjacent():
+def test_grid_qubit_is_adjacent():
```
`xmon` still on line 82.

dabacon replied:
done

cduck commented on cirq/line/line_qubit.py:
```diff
@@ -19,7 +19,17 @@
+    One can construct new LineQubits by adding or subtracting integers:
+
+        LineQubit(1) + 3 is LineQubit(4)
+        LineQubit(2) - 1 is LineQubit(1)
```
Format this for doctest also.

dabacon replied:
Done

cduck commented on cirq/line/line_qubit.py:
```diff
@@ -49,3 +59,23 @@ def __repr__(self):
+
+    def __sub__(self, other: int) -> 'LineQubit':
+        assert isinstance(other, int), (
+            'Can only subtract ints to LineQubits. Instead was {}'.format(other)
```
*from* LineQubits?

dabacon replied:
reworded (and others)

cduck commented on cirq/line/line_qubit_test.py:
```diff
@@ -83,3 +83,14 @@ def test_range():
+
+
+def test_neg():
+    assert -LineQubit(1) == LineQubit(-1)
```
Nit: Needs a newline.

dabacon replied:
done

cduck commented on cirq/line/line_qubit.py:
```diff
@@ -49,3 +62,23 @@ def __repr__(self):
+    def __sub__(self, other: int) -> 'LineQubit':
+        assert isinstance(
+            other, int), ('Can only subtract ints and LineQubits. Instead was'
+                          '{}'.format(other))
```
I missed this before but all these assertions should probably be TypeErrors.

dabacon replied:
Fixed.  Also missed that the linequbit test imported LineQubit directly so fixed this.  And I was missing a check of the type error in line qubit.
